drivers/counter/counter_sam_tc.c

```diff
@@ -0,0 +1,404 @@
+	void *user_data;
+};
+
+/* Device run time data */
+struct counter_sam_dev_data {
+	counter_top_callback_t top_cb;
+	void *top_user_data;
+
+	struct counter_sam_alarm_data alarm[MAX_ALARMS_PER_TC_CHANNEL];
+};
+
+#define DEV_NAME(dev) ((dev)->name)
+#define DEV_CFG(dev) \
+	((const struct counter_sam_dev_cfg *const)(dev)->config)
+#define DEV_DATA(dev) \
+	((struct counter_sam_dev_data *const)(dev)->data)
+
+static const uint32_t sam_tc_input_freq_table[] = {
+#if defined(CONFIG_SOC_SERIES_SAME70) || defined(CONFIG_SOC_SERIES_SAMV71)
+	USEC_PER_SEC,
+	SOC_ATMEL_SAM_MCK_FREQ_HZ / 8,
+	SOC_ATMEL_SAM_MCK_FREQ_HZ / 32,
+	SOC_ATMEL_SAM_MCK_FREQ_HZ / 128,
+	32768,
+#elif defined(CONFIG_SOC_SERIES_SAM4L)
+	USEC_PER_SEC,
+	SOC_ATMEL_SAM_MCK_FREQ_HZ / 2,
+	SOC_ATMEL_SAM_MCK_FREQ_HZ / 8,
+	SOC_ATMEL_SAM_MCK_FREQ_HZ / 32,
+	SOC_ATMEL_SAM_MCK_FREQ_HZ / 128,
+#else
+	SOC_ATMEL_SAM_MCK_FREQ_HZ / 2,
+	SOC_ATMEL_SAM_MCK_FREQ_HZ / 8,
+	SOC_ATMEL_SAM_MCK_FREQ_HZ / 32,
+	SOC_ATMEL_SAM_MCK_FREQ_HZ / 128,
+	32768,
+#endif
+	USEC_PER_SEC, USEC_PER_SEC, USEC_PER_SEC,
+};
+
+static int counter_sam_tc_start(const struct device *dev)
+{
+	const struct counter_sam_dev_cfg *const dev_cfg = DEV_CFG(dev);
+	Tc *tc = dev_cfg->regs;
+	TcChannel *tc_ch = &tc->TcChannel[dev_cfg->tc_chan_num];
+
+	tc_ch->TC_CCR = TC_CCR_CLKEN | TC_CCR_SWTRG;
+
+	return 0;
+}
+
+static int counter_sam_tc_stop(const struct device *dev)
+{
+	const struct counter_sam_dev_cfg *const dev_cfg = DEV_CFG(dev);
+	Tc *tc = dev_cfg->regs;
+	TcChannel *tc_ch = &tc->TcChannel[dev_cfg->tc_chan_num];
+
+	tc_ch->TC_CCR = TC_CCR_CLKDIS;
+
+	return 0;
+}
+
+static int counter_sam_tc_get_value(const struct device *dev, uint32_t *ticks)
+{
+	const struct counter_sam_dev_cfg *const dev_cfg = DEV_CFG(dev);
+	Tc *tc = dev_cfg->regs;
+	TcChannel *tc_ch = &tc->TcChannel[dev_cfg->tc_chan_num];
+
+	*ticks = tc_ch->TC_CV;
+
+	return 0;
+}
+
+static int counter_sam_tc_set_alarm(const struct device *dev,
+				    uint8_t chan_id,
+				    const struct counter_alarm_cfg *alarm_cfg)
+{
+	struct counter_sam_dev_data *data = DEV_DATA(dev);
+	const struct counter_sam_dev_cfg *const dev_cfg = DEV_CFG(dev);
+	Tc *tc = dev_cfg->regs;
+	TcChannel *tc_ch = &tc->TcChannel[dev_cfg->tc_chan_num];
+	uint32_t alarm_value;
+
+	__ASSERT_NO_MSG(alarm_cfg->callback != NULL);
+
+	if (chan_id >= MAX_ALARMS_PER_TC_CHANNEL) {
+		return -EINVAL;
+	}
+
+	uint32_t top_value = tc_ch->TC_RC != 0 ? tc_ch->TC_RC : UINT16_MAX + 1;
+
+	if (alarm_cfg->ticks > top_value) {
+		return -EINVAL;
+	}
+
+	if (data->alarm[chan_id].callback != NULL) {
+		return -EBUSY;
+	}
+
+	if (chan_id == 0) {
+		tc_ch->TC_IDR = TC_IDR_CPAS;
+	} else {
+		tc_ch->TC_IDR = TC_IDR_CPBS;
+	}
+
+	data->alarm[chan_id].callback = alarm_cfg->callback;
+	data->alarm[chan_id].user_data = alarm_cfg->user_data;
+
+	if ((alarm_cfg->flags & COUNTER_ALARM_CFG_ABSOLUTE) != 0) {
+		alarm_value = alarm_cfg->ticks;
+	} else {
+		alarm_value = (tc_ch->TC_CV + alarm_cfg->ticks) % top_value;
```

Comment:
there is no protection against ticks being to small (e.g. 1).

It mentions above that guard periods are not supported but they apply to absolute alarms. There is an assumption that relative alarm will always expire as expected and not after full cycle.

Reply:
@nordic-krch Thanks for the review! All issues should be fixed now apart from short relative alarms. Implementing this correctly is not trivial. For now I've added the follwing comment to the driver source file as well as the commit

> - The driver does not guarantee that short relative alarm will trigger the interrupt immediately and not after the full cycle / counter overflow.

I would prefer if the next person that needs a feature implements and tests it. Hope that's OK.